### PR TITLE
#294 Implement onehot encoding

### PR DESCRIPTION
**Pull request by Lewuathe**
see #294 
usage
```sql
SELECT 
  inline(
    onehot_encoding(code)
  )
FROM
  www_access
```

|label|encode|
|:---:|:---:|
|200|[1,0,0]|
|404|[0,1,0]|
|500|[0,0,1]|

We can join onehot encoding table with source table in order to created onehot labeled datas like,

```sql
SELECT 
  source.path,
  source.host,
  encoded.encode
FROM
  www_access source
JOIN (
    SELECT 
      inline(
        onehot_encoding(code)
      )
  FROM
    www_access
) encoded
ON encoded.label = source.code
```
|path|host|encode|
|:---:|:---:|:---|
|/path/to/1|example1.com|[1,0,0]|
|/path/to/2|example2.com|[0,1,0]|

### CODE REVIEW

**Comment by Lewuathe** [comment]
@myui scikit-learn OneHotEncoder [only supports int type value](https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/preprocessing/data.py#L1751-L1752). 

The implementation of GraphLab seems support string type and the behaviour of string type is same to int one. So indexing against sorted set seems to be same to these libraries.

But GraphLab supports dictionary output and it's reasonable also to Hivemall because one hot encoded feature must be very sparse. So I'll update to output dictionary type output.

**Comment by myui** [comment]
@Lewuathe Could you describe what kind of input/output you are expecting?

**Comment by Lewuathe** [comment]
@myui Thanks for checking. I think it's difficult to encode each samples with Hive UDF because OneHot encoding requires UDAF and UDTF features (aggregation and explosion). We can do it by `explode` or `inline` UDF as did in `histogram_numeric`. But the number of samples can be huge unlike `histogram_numeric`. It's unreasonable to explode it.

So I decided to only provide encoding table as described [here](https://github.com/myui/hivemall/pull/318#issue-167059625). We can recreated encoded samples with joining with encoding table. 

We can provide sparse format in addition to dense format. So the input/output which I'm expecting looks like

| category1 | category2 |
|:---:|:---:|
| cat | mammal |
| dog| mammal |
| human | mammal |
| seahawk | bird |
| wasp | insect| 

↓

| category1 | category2 | encoded_features | 
|:---:|:---:|:---:|
| cat | mammal | {1, 6} |
| dog| mammal | {2,6} |
| human | mammal | {3,6} |
| seahawk | bird | {4,7} |
| wasp | insect| {5,8} |

Because we need to provide original columns to join with original data.

**Comment by myui** [comment]
@Lewuathe building encoded table by UDAF sounds fine. `trasform()` would be done by SQL inner join.

**Comment by Lewuathe** [comment]
Hmm, I'm facing class cast exception at the end of reduce phase with updated code...
```
Caused by: org.apache.hadoop.hive.ql.metadata.HiveException: java.lang.ClassCastException: org.apache.hadoop.io.Text cannot be cast to [Ljava.lang.Object;
       	at org.apache.hadoop.hive.ql.exec.GroupByOperator.flush(GroupByOperator.java:1082)
       	at org.apache.hadoop.hive.ql.exec.GroupByOperator.closeOp(GroupByOperator.java:1121)
       	... 9 more
Caused by: java.lang.ClassCastException: org.apache.hadoop.io.Text cannot be cast to [Ljava.lang.Object;
       	at org.apache.hadoop.hive.serde2.objectinspector.StandardListObjectInspector.getList(StandardListObjectInspector.java:111)
       	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.serialize(LazySimpleSerDe.java:492)
       	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.serialize(LazySimpleSerDe.java:545)
       	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.serialize(LazySimpleSerDe.java:501)
```

File: core/src/main/java/hivemall/ftvec/OnehotEncodingUDAF.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object terminatePartial(AggregationBuffer aggregationBuffer) throws HiveException {
+            EncodingBuffer buf = (EncodingBuffer)aggregationBuffer;
+            List<Text> ret = new ArrayList<Text>();
+            // TODO: HIVE-11176 - Fail to passing List<Text> due to class cast exception.
+            // As workaround, string type joined under score is passed.
+            for (ComparableList<Text> features : buf.getFeaturesSet()) {
+                ret.add(new Text(StringUtils.join(features, "_")));
```
Comment (myui) [inline core/src/main/java/hivemall/ftvec/OnehotEncodingUDAF.java]: What happens if feature contains `_` on `merge()`. It would produce invalid splits.

I think it is better to have `byte[] EncodingBuffer#serialize()` method and  static`EncodingBuffer.decode(byte[] b)` method.

Comment (Lewuathe) [inline core/src/main/java/hivemall/ftvec/OnehotEncodingUDAF.java]: Thanks. I'll do so.

**Comment by Lewuathe** [comment]
```
hive> select * from records;
OK
1989   	1      	2
1990   	1      	2
2011   	1      	3
2012   	2      	3
2013   	2      	2
2014   	2      	3
hive> desc records;
OK
year                   	string
temperature            	int
quality                	int
Time taken: 0.131 seconds, Fetched: 3 row(s)
```

This table can be converted to 

```
hive> select inline(onehot_encoding(year)) from records;
OK
1989   	["0:1"]
1990   	["1:1"]
2011   	["2:1"]
2012   	["3:1"]
2013   	["4:1"]
2014   	["5:1"]

hive> select inline(onehot_encoding(temperature)) from records;
OK
1      	["0:1"]
2      	["1:1"]
```

One thing I found is there is no way to pass always number of features to `init` method properly because `init` method is called with *default constructor in FINAL or COMPLETE* phase. So it seems there is no good way to inform UDAF Evaluator the number of features.
Therefore this implementation [fixed the number of feature as 1](https://github.com/myui/hivemall/pull/318/files#diff-899c728413caeaab2428748331ee6389R107).

**Comment by Lewuathe** [comment]
@myui Could you check this when you get a chance?

> One thing I found is there is no way to pass always number of features to init method properly because init method is called with default constructor in FINAL or COMPLETE phase.

And please let me know if you know any workaround about passing number of features.

**Comment by myui** [comment]
Sure! I'll search for workaround.

**Comment by myui** [comment]
merged in #357
